### PR TITLE
feat: (Uni-commerce) generate Delivery Note and sync item fields

### PR DESCRIPTION
**Pull request by Vikas8600**
feat: (Uni-commerce) generate Delivery Note and sync item fields

We have implemented the requested functionality as per the client's specifications. 
Specifically, when the designated checkbox is clicked, the system will automatically import Delivery Notes from Uni-commerce into ERPNEXT upon shipment.

image

https://komododecks.com/recordings/uXlkZvS50jHgA3oxK98y

### CODE REVIEW

File: ecommerce_integrations/unicommerce/order.py [hunk 1]
```diff
@@ -3,386 +3,403 @@
...
+    IS_COD_CHECKBOX,
+    MODULE_NAME,
+    ORDER_CODE_FIELD,
+    ORDER_ITEM_BATCH_NO,
+    ORDER_ITEM_CODE_FIELD,
+    ORDER_STATUS_FIELD,
+    PACKAGE_TYPE_FIELD,
+    SETTINGS_DOCTYPE,
+    TAX_FIELDS_MAPPING,
+    TAX_RATE_FIELDS_MAPPING,
```
Comment (ankush) [inline ecommerce_integrations/unicommerce/order.py]: can you undo all these formatting changes, quote hard to find what actually changed here.

File: ecommerce_integrations/unicommerce/custom/delivery_note.json [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+   "columns": 0,
+   "creation": "2023-04-07 21:02:25.665457",
+   "default": null,
+   "depends_on": null,
+   "description": null,
+   "docstatus": 0,
+   "dt": "Delivery Note",
+   "fetch_from": null,
+   "fetch_if_empty": 0,
+   "fieldname": "unicommerce_order_no",
```
Comment (ankush) [inline ecommerce_integrations/unicommerce/custom/delivery_note.json:82]: Custom fields should be managed by unicommerce setting's `setup_custom_fields` function. 


Fixutres are auto imported which shouldn't be the case for integrations which are not enabled by default.

Comment (ankush) [reply ecommerce_integrations/unicommerce/custom/delivery_note.json:82]: Check this

https://github.com/frappe/ecommerce_integrations/blob/3cf612bd468cec39fc7fe722c3826226245002cf/ecommerce_integrations/unicommerce/doctype/unicommerce_settings/unicommerce_settings.py#L182

File: ecommerce_integrations/unicommerce/customer.py [hunk 3]
```diff
@@ -117,5 +117,6 @@ def _create_customer_address(uni_address, address_type, customer, also_shipping=
 			"links": [{"link_doctype": "Customer", "link_name": customer.name}],
 			"is_primary_address": int(address_type == "Billing"),
 			"is_shipping_address": int(also_shipping or address_type == "Shipping"),
+			"gst_category": 'Overseas' if country != 'India' else 'Unregistered'
```
Comment (ankush) [inline ecommerce_integrations/unicommerce/customer.py]: This shouldn't be assumed always that it's `Unregistered` if indian? 

Maybe leave this as customization on site?

File: ecommerce_integrations/unicommerce/delivery_note.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    "RETURN_ACKNOWLEDGED",
+    "MERGED",
+]
+
+import time
+
+@frappe.whitelist()
+def prepare_delivery_note():
+    try:
+        time.sleep(15)
```
Comment (ankush) [inline ecommerce_integrations/unicommerce/delivery_note.py]: what is this?

File: ecommerce_integrations/unicommerce/invoice.py [hunk 5]
```diff
@@ -381,8 +381,9 @@ def create_sales_invoice(
 	si.naming_series = channel_config.sales_invoice_series or settings.sales_invoice_series
 	si.delivery_date = so.delivery_date
 	si.ignore_pricing_rule = 1
-	si.update_stock = update_stock
+	# si.update_stock = update_stock
```
Comment (ankush) [inline ecommerce_integrations/unicommerce/invoice.py:384]: Why is this disabled? this is how other users use it. 

Don't just change code to achieve what you want, the integration is used by many users differently. You have to make everything configurable and default behaviour should continue to be there. 

Why you need delivery notes separately anyway?

File: ecommerce_integrations/unicommerce/invoice.py [hunk 6]
```diff
@@ -381,8 +381,9 @@ def create_sales_invoice(
...
 	si.flags.raw_data = si_data
+	si.debit_to = frappe.get_value("Account",{"account_currency":so.currency},"name")
```
Comment (ankush) [inline ecommerce_integrations/unicommerce/invoice.py]: This makes no sense. It will fetch any account which has currency same as sales order.

File: ecommerce_integrations/unicommerce/delivery_note.py (hunk 4 above)
Comment (ankush) [reply ecommerce_integrations/unicommerce/delivery_note.py]: can you summarize in PR why delivery notes are generated and why it can't be done with default behaviour?

File: ecommerce_integrations/unicommerce/order.py [hunk 7]
```diff
@@ -386,3 +395,14 @@ def _get_warehouse_allocations(sales_order):
...
+
+@frappe.whitelist()
+def resync_item(code):
+	from ecommerce_integrations.unicommerce.utils import force_sync
+
+	id_ecommerce = frappe.get_value("Ecommerce Item", {"erpnext_item_code": code})
+	if id_ecommerce:
+		frappe.delete_doc("Ecommerce Item", id_ecommerce)
+		force_sync(document="Items")
+	return
```
Comment (ankush) [inline ecommerce_integrations/unicommerce/order.py:408]: If it's about ecommerce item it should be in `product.py`?

Comment (ankush) [reply ecommerce_integrations/unicommerce/order.py:408]: The item isn't actually deleted from unicommerce so how does resync fix anything?

File: ecommerce_integrations/unicommerce/order.py [hunk 8]
```diff
@@ -78,8 +87,9 @@ def _get_new_orders(
 	for order in uni_orders:
 		if order["channel"] not in configured_channels:
 			continue
-		if frappe.db.exists("Sales Order", {ORDER_CODE_FIELD: order["code"]}):
-			continue
+		if check:
```
Comment (ankush) [inline ecommerce_integrations/unicommerce/order.py:90]: Why is this added if it's never set to `False` anywhere?

File: ecommerce_integrations/unicommerce/order.py [hunk 9]
```diff
@@ -48,25 +48,34 @@ def sync_new_orders(client: UnicommerceAPIClient = None, force=False):
...
-	new_orders = _get_new_orders(client, status=status)
-
+	new_orders = _get_new_orders(client, status=status, check=True)
 	if new_orders is None:
 		return
-
 	for order in new_orders:
 		sales_order = create_order(order, client=client)
+		# if settings.only_sync_completed_orders:
+		#     _create_sales_invoices(order, sales_order, client)
```
Comment (ankush) [inline ecommerce_integrations/unicommerce/order.py:57]: Why is this just removed? Again, don't just break existing things because multiple people use it.

**Review by ankush — CHANGES_REQUESTED** [review]
This looks very premature change right now. First please update the description first to describe what are the changes made here. 

Plenty of changes appear to be customizations needed for a particular user, such changes:
1. It shouldn't be in the core, should be handled via hooks/customizations. 
2. If required should be behind configuration in Unicommerce settings
3. Should not break default behaviour (breaking changes)

Also, this PR is breaking tests, so pretty sure some default functionalities broke because of it. You can check the GitHub action run for tests that are failing.

File: ecommerce_integrations/unicommerce/product.py [hunk 10]
```diff
@@ -287,7 +286,11 @@ def _build_unicommerce_item(item_code: ItemCode) -> JsonDict:
 	)
 	# append site prefix to image url
 	item_json["imageUrl"] = get_url(item.image)
-
+	item_json["maxRetailPrice"] = frappe.get_value("Item Price",{"item_code":item_code},"price_list_rate")
+	item_json["hsnCode"] = frappe.get_value("Item",item.name,"gst_hsn_code")
+	item_json["description"] = frappe.utils.strip_html_tags(frappe.get_value("Item",item.name,"description"))
+	item_json["gstTaxTypeCode"] = frappe.get_value("Item Tax",{"parent":item.name},"item_tax_template")
```
[the remark this answers is not in the document]
Comment (Vikas8600) [reply ecommerce_integrations/unicommerce/product.py]: Removed

File: ecommerce_integrations/unicommerce/order.py (hunk 1 above)
Comment (Vikas8600) [reply ecommerce_integrations/unicommerce/order.py]: Reverted

File: ecommerce_integrations/unicommerce/custom/delivery_note.json (hunk 2 above)
Comment (Vikas8600) [reply ecommerce_integrations/unicommerce/custom/delivery_note.json]: Added

File: ecommerce_integrations/unicommerce/customer.py (hunk 3 above)
Comment (Vikas8600) [reply ecommerce_integrations/unicommerce/customer.py]: Removed

File: ecommerce_integrations/unicommerce/delivery_note.py (hunk 4 above)
Comment (Vikas8600) [reply ecommerce_integrations/unicommerce/delivery_note.py]: Removed

File: ecommerce_integrations/unicommerce/invoice.py (hunk 5 above)
Comment (Vikas8600) [reply ecommerce_integrations/unicommerce/invoice.py]: No prior delivery note had been created,
and the stock is directly updated on the sale invoice

Now it's configure with Delivery note settings

File: ecommerce_integrations/unicommerce/invoice.py (hunk 6 above)
Comment (Vikas8600) [reply ecommerce_integrations/unicommerce/invoice.py]: Removed

File: ecommerce_integrations/unicommerce/order.py (hunk 7 above)
Comment (Vikas8600) [reply ecommerce_integrations/unicommerce/order.py]: Removed

File: ecommerce_integrations/unicommerce/order.py (hunk 8 above)
Comment (Vikas8600) [reply ecommerce_integrations/unicommerce/order.py]: Removed

File: ecommerce_integrations/unicommerce/order.py (hunk 9 above)
Comment (Vikas8600) [reply ecommerce_integrations/unicommerce/order.py]: Reverted

Comment (Vikas8600) [reply ecommerce_integrations/unicommerce/order.py]: I've added test cases, it would be great if you could go over them.

**Comment by Vikas8600** [comment]
I've added test cases; it would be great if you could go over them.
